File: Source/Test/Yamo.Test/Model/PropertyModifiedTrackingBase.vb
﻿Namespace Model

  Public MustInherit Class PropertyModifiedTrackingBase
    Implements IHasDbPropertyModifiedTracking

    ' NOTE: this is very naive and memory consuming implementation of IHasDbPropertyModifiedTracking

    Private m_Modified As HashSet(Of String) = New HashSet(Of String)

    Protected Sub MarkPropertyAsModified(propertyName As String)
      m_Modified.Add(propertyName)
    End Sub

    Public Function IsAnyDbPropertyModified() As Boolean Implements IHasDbPropertyModifiedTracking.IsAnyDbPropertyModified
      Return m_Modified.Any()
    End Function

    Public Function IsDbPropertyModified(propertyName As String) As Boolean Implements IHasDbPropertyModifiedTracking.IsDbPropertyModified
      Return m_Modified.Contains(propertyName)
    End Function

    Public Sub ResetDbPropertyModifiedTracking() Implements IHasDbPropertyModifiedTracking.ResetDbPropertyModifiedTracking
      m_Modified.Clear()
    End Sub

  End Class
End Namespace

File: Source/Samples/Yamo.PlaygroundCS/Model/PropertyModifiedTrackingBase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Yamo.PlaygroundCS.Model
{
    public abstract class PropertyModifiedTrackingBase : IHasDbPropertyModifiedTracking
    {

        //  NOTE: this is very naive and memory consuming implementation of IHasDbPropertyModifiedTracking
        private HashSet<string> m_Modified = new HashSet<string>();

        protected void MarkPropertyAsModified(string propertyName)
        {
            m_Modified.Add(propertyName);
        }

        public bool IsAnyDbPropertyModified()
        {
            return m_Modified.Any();
        }

        public bool IsDbPropertyModified(string propertyName)
        {
            return m_Modified.Contains(propertyName);
        }

        public void ResetDbPropertyModifiedTracking()
        {
            m_Modified.Clear();
        }
    }
}
